### ISSUE TITLE
Build fails with `Error: Cannot find module 'html-webpack-plugin'`

### ISSUE THREAD

**Issue by gturri** [vivreanantes/cartovrac#20]
Hi,
I'm following the documentation https://vivreanantes.github.io/cartovrac/how_to_build_cartovrac.html but I'm having trouble to build this project.

**In a nutshell**

Here is what I'm doing:

    git clone github.com/vivreanantes/cartovrac/
    cd cartovrac
    npm install --save-dev webpack
    npm run build

But the last commands fails with `Error: Cannot find module 'html-webpack-plugin'`.

I'm afraid I'm not very familiar with node so I don't know if it's related to something I'm doing wrong. I would be thrilled if you can help understand what I'm missing

**More details**

I'm using ubuntu 18.04 and I just installed the latest node version from https://nodejs.org/en/ (`npm --version` => `6.13.4`)

The `npm install` command is successful but outputs a few warnings:

    npm notice created a lockfile as package-lock.json. You should commit this file.
    npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.11 (node_modules/fsevents):
    npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.11: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

The `npm run build` command starts by telling me

    One CLI for webpack must be installed. These are recommended choices, delivered as separate packages:
     - webpack-cli (https://github.com/webpack/webpack-cli)
       The original webpack full-featured CLI.
    We will use "npm" to install the CLI via "npm install -D".
    Do you want to install 'webpack-cli' (yes/no): 

I enter `yes` and then it fails with this log:

    Installing 'webpack-cli' (running 'npm install -D webpack-cli')...
    npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.11 (node_modules/fsevents):
    npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.11: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
    
    + webpack-cli@3.3.10
    added 64 packages from 30 contributors and audited 5288 packages in 8.398s
    
    2 packages are looking for funding
      run `npm fund` for details
    
    found 0 vulnerabilities
    
    Error: Cannot find module 'html-webpack-plugin'
    Require stack:
    - /documents/zwp/cartovrac/webpack.config.js
    - /documents/zwp/cartovrac/node_modules/webpack-cli/bin/utils/convert-argv.js
    - /documents/zwp/cartovrac/node_modules/webpack-cli/bin/cli.js
    - /documents/zwp/cartovrac/node_modules/webpack/bin/webpack.js
        at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)
    ...huge stacktrace skipped for brevity...
        at /documents/zwp/cartovrac/node_modules/webpack/bin/webpack.js:143:5
        at processTicksAndRejections (internal/process/task_queues.js:93:5) {
      code: 'MODULE_NOT_FOUND',
      requireStack: [
        '/documents/zwp/cartovrac/webpack.config.js',
        '/documents/zwp/cartovrac/node_modules/webpack-cli/bin/utils/convert-argv.js',
        '/documents/zwp/cartovrac/node_modules/webpack-cli/bin/cli.js',
        '/documents/zwp/cartovrac/node_modules/webpack/bin/webpack.js'
      ]
    }
    npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! cartovrac@1.0.0 build: `webpack --mode production;`
    npm ERR! Exit status 1
    npm ERR! 
    npm ERR! Failed at the cartovrac@1.0.0 build script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/guillaume/.npm/_logs/2019-12-22T22_05_33_920Z-debug.log

**Comment by gturri** [comment]
To add some more information: it turns out I can get rid of this error by running `npm install html-webpack-plugin` (plus several other `npm install` for all the other `Cannot find module xxx` I get afterwards)

However `npm run build` eventually fails with this other kind of error

```
> cartovrac@1.0.0 build /documents/zwp/cartovrac
> webpack --mode production;

Executing pre-build scripts
Successfully wrote file
Successfully wrote file
Successfully wrote file
(node:24356) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
Hash: fcadbb24f1132c249090
Version: webpack 4.41.4
Time: 74767ms
Built at: 12/24/2019 5:46:09 PM
 1 asset
Entrypoint main = main.js
 [0] ./src/config.js 797 bytes {0} [built]
 [1] ./src/data.js 4.16 KiB {0} [built]
 [2] ./src/map.js 3.03 KiB {0} [built]
 [3] ./src/popup.js 3.28 KiB {0} [built]
 [4] ./src/osm_eco_extractor.js 7.65 KiB {0} [built]
 [5] ./src/index.js 1.94 KiB {0} [built]
 [6] ./src/menu.js 756 bytes {0} [built]
 [7] ./cache_jtb_data_simplified.json 20.2 KiB {0} [built]
 [8] ./cache_cyclad_data_simplified.json 12.5 KiB {0} [built]
 [9] ./cache_bulk_data_simplified.json 342 KiB {0} [built]
[10] ./data/bulk/itinerant.json 12.4 KiB {0} [built]
[11] ./src/leaflet.groupedlayercontrol.js 9.84 KiB {0} [built]
[14] (webpack)/buildin/module.js 497 bytes {0} [built]
[15] (webpack)/buildin/global.js 472 bytes {0} [built]
    + 6 hidden modules

ERROR in ./src/index.js
Module not found: Error: Can't resolve 'jquery' in '/documents/zwp/cartovrac/src'
 @ ./src/index.js 1:0-17

ERROR in ./src/menu.js
Module not found: Error: Can't resolve 'jquery' in '/documents/zwp/cartovrac/src'
 @ ./src/menu.js 1:0-17
 @ ./src/index.js

ERROR in ./src/osm_eco_extractor.js
Module not found: Error: Can't resolve 'leaflet' in '/documents/zwp/cartovrac/src'
 @ ./src/osm_eco_extractor.js 1:0-18
 @ ./src/index.js

// etc... lot of other "module not found: Error: Can't resolve 'xxx' errors
```

so I'm not sure if I'm progressing in the good direction here.
(also, since those `npm install` aren't mentioned in the doc, I'm not sure I've done the right thing)

**Comment by gturri** [comment]
For the record: I manage to have a successful `npm run build` when I install a whole bunch of dependencies:

       npm install html-webpack-plugin
       npm install extract-text-webpack-plugin
       npm install webpack-synchronizable-shell-plugin
       npm install jquery
       npm install leaflet
       npm install jquery
       npm install leaflet
       npm install leaflet-control-geocoder
       npm install mapbox-gl-leaflet
       npm i style-loader
       npm i typeface-roboto
       npm i url-loader
       npm i html-loader
       npm i file-loader
       npm i css-loader
       npm i leaflet.featuregroup.subgroup
       npm i leaflet.markercluster
       npm i https://github.com/domoritz/leaflet-locatecontrol
       npm i mapbox-gl

I'm not sure why I needed all this and I'm probably doing something wrong. I would be glad to understand what because it might be painful if I have to do that each time I clone this repo on another dev station

**Comment by AntoineJaury** [comment]
Hi @gturri, you shouldn't have to install these dependencies manually. I'm trying to replicate your issue to better understand why it's not working. I'll come back soon with a fix on the configuration or a workaround in the doc.

**Comment by AntoineJaury** [comment]
Ok, the problem you have is due to this command `npm install --save-dev webpack` which only install `devDependencies`. The configuration of the dependencies I made is wrong. Some dependencies used by webpack (the one missing when you launched the build) were defined as normal dependencies instead of dev one.

I fixed the dependency definition: https://github.com/vivreanantes/cartovrac/commit/472eecc1f7810977155789674a2ffe4147f1a146

Now, running same commands as you did following the tutorial should work:

```
git clone github.com/vivreanantes/cartovrac/
cd cartovrac
npm install --save-dev webpack
npm run build
```
Feel free to re-open the issue if you meet this case again.

**Comment by gturri** [comment]
Hi,

thanks for your feedback. However I'm afraid it doesn't fix the issue on my side. I've just started from a fresh new clone of cartovrac (I'm on the current master:  9c280a1e4201a62c126df09676e48198a56a4a75 ; which hence includes the commit mentioned above)

From this clone I ran

    npm install --save-dev webpack
    npm run build # I accept when this command interactively ask me to install webpack-cli

but it fails with the same error:

    Error: Cannot find module 'html-webpack-plugin'

FTR, my config is:
* Ubuntu 18.04.3 LTS
* npm --version => 3.5.2

(I'm afraid I don't have the permissions to re-open this issue. Let me know if you want me to create a new one instead)
